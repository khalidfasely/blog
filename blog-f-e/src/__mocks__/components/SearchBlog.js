//function SearchBlog({ children }){
//    return <>{children}</>;
//  }
//  
//  export default SearchBlog;

export default jest.mock('../../components/SearchBlog', () => () => 'SearchBlog');
