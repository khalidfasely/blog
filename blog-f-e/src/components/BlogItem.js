import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import ReactMarkdown from "react-markdown";
import { startDeleteBlog } from '../actions/blogs';

const BlogItem = ({ blog, uname, startDeleteBlog }) => {
  const [rModalOpen, setRModalOpen] = useState(false);
  const [eModalOpen, setEModalOpen] = useState(false);

  const deleteBlog = () => {
    startDeleteBlog(blog.id);
    setRModalOpen(false);
  }

  return (
    <div>
      <Link to={`/blog/${blog.id}`}>{blog.title}</Link> by <Link to={`/user/${blog.created_by.id}`}>{blog.created_by.username}</Link> -- {blog.created_at}
      <div>{blog.description}</div>
      <div>{blog.likes} -- {blog.dislikes}</div>
      {
        uname===blog.created_by.username &&
        <div>
          <button onClick={() => setRModalOpen(true)}>X</button>
          <button onClick={() => setEModalOpen(true)}>Edit</button>
        </div>
      }
      <Modal
        isOpen={rModalOpen}
        contentLabel="Delete Blog"
        closeTimeoutMS={100}
      >
        <h5>If you delete this blog there is no way to return it.</h5>
        <button onClick={deleteBlog}>Delete Blog</button>
        <button onClick={() => setRModalOpen(false)}>X</button>
      </Modal>
    </div>
  )
};//closeTimeoutMS={200}

const mapStateToProps = (state) => ({
  uname: state.auth.uname
});

const mapDispatchToProps = (dispatch) => ({
  startDeleteBlog: (bid) => dispatch(startDeleteBlog(bid))
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogItem);

//<ReactMarkdown>{blog.content}</ReactMarkdown>