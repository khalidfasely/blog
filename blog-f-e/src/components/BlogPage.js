import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import ReactMarkdown from 'react-markdown';
import { startAddComment, startLikeBlog, startSetBlog, startUnlikeBlog } from '../actions/blogPage';
import Comment from './Comment';
import { Link } from 'react-router-dom';
import { l_blog, s_blog, u_blog, u_s_blog } from '../actions/auth';
import { startSaveBlog, startUnsaveBlog } from '../actions/savedBlogs';
import ModalEdit from './ModalEdit';
import previewContent from '../functions/previewContent';

const BlogPage = (props) => {
    const [renderBlog, setRenderBlog] = useState(false);
    const [blog, setBlog] = useState(undefined);
    const [newComment, setNewComment] = useState('');
    const [ buttonDis, setButtonDis ] = useState(false);
    const [ sButtonDis, setSButtonDis ] = useState(false);
    const [eModalOpen, setEModalOpen] = useState(false);
    const [ previewBlog, setPreviewBlog ] = useState(undefined);
    const disableButton = () => {
        setButtonDis(true);
        setTimeout(() => {
            setButtonDis(false);
        }, 200);
    };
    const disableSButton = () => {
        setSButtonDis(true);
        setTimeout(() => {
            setSButtonDis(false);
        }, 200);
    };
    const blogDidAlreadyLoad = () => {
        let blogInList = undefined;
        props.blogList.map((blogItem) => {
            if(blogItem.id === parseInt(props.match.params.bid)){
                blogInList = blogItem;
            };
        });
        return blogInList;
    }
    useEffect(() => {
        const blog = blogDidAlreadyLoad();
        if (blog) {
            setBlog(blog);
            document.title = `${blog.title} (@${blog.created_by.username}) - Blog`;
            setRenderBlog(true);
            setPreviewBlog(previewContent(blog.content.substring(0, 300)));
            // Render the blog content.. from Redux
            // Otherwise Fecth it from the backend
            // Store it on Redux
            // Then render the content from Redux
        } else {
            props.startSetBlog(parseInt(props.match.params.bid)).then((result) => {
                setBlog({
                    ...result.blog,
                    comments: result.comments
                });
                document.title = `${result.blog.title} (@${result.blog.created_by.username}) - Blog`;
                setRenderBlog(true);
                setPreviewBlog(previewContent(result.blog.content.substring(0, 300)));
                //this.setState({ renderBlog: true });
            });
        }
        //return () => {
        //    setPreviewBlog(undefined);
        //}
    }, [])
    const onCommentChange = (e) => {
        const newComment = e.target.value;
        setNewComment(newComment);
    }

    const onFormSubmit = (e) => {
        e.preventDefault();
        if (newComment.replace(/\s/g, '')) {    
            props.startAddComment(parseInt(props.match.params.bid), newComment)//.then((result) => {

//            })
            setNewComment('');
        }
    }

    const like_b = () => {
        disableButton();
        props.startLikeBlog(parseInt(props.match.params.bid)).then(() => {
            props.l_blog(parseInt(props.match.params.bid));
            setBlog({
                ...blog,
                likes: blog.likes + 1
            });
        });
    };

    const unlike_b = () => {
        disableButton();
        props.startUnlikeBlog(parseInt(props.match.params.bid)).then(() => {
            props.u_blog(parseInt(props.match.params.bid));
            setBlog({
                ...blog,
                likes: blog.likes - 1
            });
        });
    };

    const save_b = () => {
        disableSButton();
        props.startSaveBlog(parseInt(props.match.params.bid)).then(() => {
            props.s_blog(parseInt(props.match.params.bid));
        });
    };

    const unsave_b = () => {
        disableSButton();
        props.startUnsaveBlog(parseInt(props.match.params.bid)).then(() => {
            props.u_s_blog(parseInt(props.match.params.bid));
        });
    };

    if(renderBlog) {
        if (blog.isDefault) {
            return <div>This Blog Not Available Now. <Link to='/'>Back Home</Link></div>
        } else if (props.isPreview) {
            return (
                <div>
                    <h1>{blog.title}</h1>
                    <p>{blog.created_at}</p>
                    <p>{blog.created_by.username}</p>
                    <p>{blog.category}</p>
                    <p>{blog.description}</p>
                    <ReactMarkdown>{`${previewBlog}...`}</ReactMarkdown>
                    <Link to={`/blog/${blog.id}`}>See More</Link>
                    <p>{blog.likes}</p>
                    <p>{blog.dislikes}</p>
                    <div>
                        {
                            props.uname && (
                                props.blogsLiked.includes(props.match.params.bid) ?
                                <button disabled={buttonDis} onClick={unlike_b}>Unlike</button> :
                                <button disabled={buttonDis} onClick={like_b}>Like</button>
                                )
                        }
                    </div>
                    <div>
                        {
                            props.uname && (
                                props.blogsSaved.includes(props.match.params.bid) ?
                                <button disabled={sButtonDis} onClick={unsave_b}>Unsave</button> :
                                <button disabled={sButtonDis} onClick={save_b}>Save</button>
                            )
                        }
                    </div>
                    <div>{blog.comments.length} Comment(s)</div>
                </div>
            )
        }
        return (
            <div>
                <h1>{blog.title}</h1>
                <p>{blog.created_at}</p>
                <p>{blog.created_by.username}</p>
                <p>{blog.category}</p>
                <p>{blog.description}</p>
                <ReactMarkdown>{blog.content}</ReactMarkdown>
                <p>{blog.likes}</p>
                <p>{blog.dislikes}</p>
                {
                    props.uname && (
                        props.uname === blog.created_by.username &&
                        <button onClick={() => setEModalOpen(true)}>Edit</button>
                    )
                }
                <ModalEdit eModalOpen={eModalOpen} blog={blog} setEModalOpen={setEModalOpen} />
                <div>
                    {
                        props.uname && (
                            props.blogsLiked.includes(props.match.params.bid) ?
                            <button disabled={buttonDis} onClick={unlike_b}>Unlike</button> :
                            <button disabled={buttonDis} onClick={like_b}>Like</button>
                        )
                    }
                </div>
                <div>
                    {
                        props.uname && (
                            props.blogsSaved.includes(props.match.params.bid) ?
                            <button disabled={sButtonDis} onClick={unsave_b}>Unsave</button> :
                            <button disabled={sButtonDis} onClick={save_b}>Save</button>
                        )
                    }
                </div>
                <div>
                    <h5>Comments:</h5>
                    {
                        blog.comments.length !== 0 ?
                        blog.comments.map(comment => <Comment key={comment.id} blogId={blog.id} comment={comment} />) :
                        <div>-NO COMMENTS-</div>
                    }
                </div>
                <div>
                    <h5>Add comments:</h5>
                    {
                        props.uname ?
                        <form onSubmit={onFormSubmit}>
                            <textarea
                            placeholder='Comment'
                            onChange={onCommentChange}
                            value={newComment}
                            />
                            <button disabled={!newComment.replace(/\s/g, '')}>Add Comment</button>
                        </form> :
                        <h5><Link to='/login'>Login</Link> or <Link to='/register'>Sign In</Link> to Add a Comment!</h5>
                    }
                </div>
            </div>
        )
    } else {
        return <div>Blog Page {props.match.params.bid}</div>
    };
};

const mapStateToProps = (state) => ({
    blogList: state.blogPage,
    uname: state.auth.uname,
    blogsLiked: state.auth.blogsLiked,
    blogsSaved: state.auth.blogsSaved
});

const mapDispatchToProps = (dispatch) => ({
    startSetBlog: (id) => dispatch(startSetBlog(id)),
    startAddComment: (blog_id, comment) => dispatch(startAddComment({ blog_id, comment })),
    startLikeBlog: (id) => dispatch(startLikeBlog(id)),
    startUnlikeBlog: (id) => dispatch(startUnlikeBlog(id)),
    l_blog: (id) => dispatch(l_blog(id)),
    u_blog: (id) => dispatch(u_blog(id)),
    s_blog: (id) => dispatch(s_blog(id)),
    u_s_blog: (id) => dispatch(u_s_blog(id)),
    startSaveBlog: (id) => dispatch(startSaveBlog(id)),
    startUnsaveBlog: (id) => dispatch(startUnsaveBlog(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogPage);

//import React, { useEffect, useState } from 'react';
//import { connect } from 'react-redux';
//import ReactMarkdown from 'react-markdown';
//import { startAddComment, startSetBlog } from '../actions/blogPage';
//import Comment from './Comment';
//
//const BlogPage = (props) => {
//    const [renderBlog, setRenderBlog] = useState(false);
//    const [blog, setBlog] = useState(undefined);
//    const [newComment, setNewComment] = useState('');
//
//    const blogDidAlreadyLoad = () => {
//        let blogInList = undefined;
//        props.blogList.map((blogItem) => {
//            if(blogItem.id === parseInt(props.match.params.bid)){
//                blogInList = blogItem;
//            };
//        });
//        return blogInList;
//    }
//    useEffect(() => {
//        const blog = blogDidAlreadyLoad();
//        if (blog) {
//            setBlog(blog);
//            setRenderBlog(true);
//            // Render the blog content.. from Redux
//            // Otherwise Fecth it from the backend
//            // Store it on Redux
//            // Then render the content from Redux
//        } else {
//            props.startSetBlog(parseInt(props.match.params.bid)).then((result) => {
//                setBlog({
//                    ...result.blog,
//                    comments: result.comments
//                });
//                setRenderBlog(true);
//                //this.setState({ renderBlog: true });
//            });
//        }
//    }, [])
//
//    const onCommentChange = (e) => {
//        const newComment = e.target.value;
//        setNewComment(newComment);
//    }
//
//    const onFormSubmit = (e) => {
//        e.preventDefault();
//        const id = parseInt(props.match.params.bid);
//        if (newComment.replace(/\s/g, '')) {    
//            props.startAddComment(id, newComment)//.then((result) => {
//
////            })
//            setNewComment('');
//        }
//    }
//
//    if(renderBlog) {
//        return (
//            <div>
//                <h1>{blog.title}</h1>
//                <p>{blog.created_at}</p>
//                <p>{blog.created_by.username}</p>
//                <p>{blog.category}</p>
//                <p>{blog.description}</p>
//                <ReactMarkdown>{blog.content}</ReactMarkdown>
//                <p>{blog.likes}</p>
//                <p>{blog.dislikes}</p>
//                <div>
//                    <h5>Comments:</h5>
//                    {
//                        blog.comments.length !== 0 ?
//                        blog.comments.map(comment => <Comment key={comment.id} comment={comment} />) :
//                        <div>-NO COMMENTS-</div>
//                    }
//                </div>
//                <div>
//                    <h5>Add comments:</h5>
//                    <form onSubmit={onFormSubmit}>
//                        <textarea
//                          placeholder='Comment'
//                          onChange={onCommentChange}
//                          value={newComment}
//                        />
//                        <button disabled={!newComment.replace(/\s/g, '')}>Add Comment</button>
//                    </form>
//                </div>
//            </div>
//        )
//    } else {
//        return <div>Blog Page {props.match.params.bid}</div>
//    };
//};
//
//const mapStateToProps = (state) => ({
//    blogList: state.blogPage
//});
//
//const mapDispatchToProps = (dispatch) => ({
//    startSetBlog: (id) => dispatch(startSetBlog(id)),
//    startAddComment: (blog_id, comment) => dispatch(startAddComment({ blog_id, comment }))
//});
//
//export default connect(mapStateToProps, mapDispatchToProps)(BlogPage);