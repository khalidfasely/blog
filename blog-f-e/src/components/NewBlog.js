import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { startEditBlog } from '../actions/blogPage';
import { editBlogFromBP, startAddBlog } from '../actions/blogs';
import { editBlogFromSB } from '../actions/savedBlogs';
import { addBlogToProfile, editBlogFromUP } from '../actions/userPage';
import { history } from '../router/AppRouter';

export const NewBlog = ({
        uname,
        profileList,
        startAddBlog,
        addBlogToProfile,
        isEdit,
        blog,
        edit,
        startEditBlog,
        editBlogFromBP,
        savedBlogs,
        editBlogFromSB,
        editBlogFromUP,
        categories
    }) => {
    //const titleDefault = blog ? blog.title : '';
    //const [ title, setTitle ] = useState(titleDefault);
    //const descriptionDefault = blog ? blog.description : '';
    //const [ description, setDescription ] = useState(descriptionDefault);
    //const contentDefault = blog ? blog.content : '';
    //const [ content, setContent ] = useState(contentDefault);
    //const categoryDefault = blog ? blog.category : '';
    //const [ category, setCategory ] = useState(categoryDefault);
    const [ title, setTitle ] = useState(blog ? blog.title : '');
    const [ description, setDescription ] = useState(blog ? blog.description : '');
    const [ content, setContent ] = useState(blog ? blog.content : '');
    const [ category, setCategory ] = useState(blog ? blog.category : '');
    const [ error, setError ] = useState('');
    const [ titleError, setTitleError ] = useState('');
    const [ descriptionError, setDescriptionError ] = useState('');
    const [ contentError, setContentError ] = useState('');
    const [ categoryError, setCategoryError ] = useState('');

    useEffect(() => {
        if(!isEdit){
            document.title = 'Create a Blog - Blog';
        }
    }, [])

    const onTitleChange = (e) => {
        const title = e.target.value;
        setTitle(title);
    }

    const onDescriptionChange = (e) => {
        const description = e.target.value;
        setDescription(description);
    }

    const onContentChange = (e) => {
        const content = e.target.value;
        setContent(content);
    }

    const onCategoryChange = (e) => {
        const category = e.target.value;
        setCategory(category);
    }

    const onFormSubmit = (e) => {
        e.preventDefault();

        setTitleError('');
        setDescriptionError('');
        setContentError('');
        setCategoryError('');

        //const availabelData = (title.replace(/\s/g, '') && description.replace(/\s/g, '') && content.replace(/\s/g, '') && category);
        const availabelTitle = title.replace(/\s/g, '').length >= 90;
        const availabelDescription = description.replace(/\s/g, '').length >= 200;
        const availabelContent = content.replace(/\s/g, '').length >= 500;
        const availabelData = (availabelTitle && availabelDescription && availabelContent && category);

        if (availabelData && !isEdit) {
            startAddBlog({ title, description, content, category }).then((result) => {
                if (result.message !== "Blog Created") {
                    setError(result.message);
                } else {
                    setError('');
                    profileList.map(profileItem => {
                        if (profileItem.uid.username === uname) {
                            // Create addBlogToProfile on userPage Action
                            addBlogToProfile(result.blog, profileItem.uid.id);
                        } else {
                            //console.log(false);
                        }
                    });
                    history.push('/');
                }
            });
        } else if (availabelData && isEdit) {
            setError('');
            //Edit It On blogPageReducer
            startEditBlog(blog.id, {uname, title, description, content, category})
            .then(() => {
                history.push('/');
                history.push(`/blog/${blog.id}`);
            })
            .then(() => {
                //Edit It On blogPageReducer
                editBlogFromBP(blog.id, {title, description, category});
            })
            .then(() => {
                //Check if blog is on savedBlogs
                savedBlogs.map(blogS => {
                    if (blogS.id === blog.id) {
                        editBlogFromSB(blog.id, {title, description, category});
                    }
                })
            })
            .then(() => {
                //Check if blog.created_by is on profileList
                profileList.map(profileItem => {
                    const profileUsername = profileItem.uid.username;
                    if (profileUsername === uname) {
                        editBlogFromUP({uname, bid: blog.id}, {title, description, category});
                    }
                })
            });
        } else {
            if(!availabelTitle){
                setTitleError('The title\'s length should be 90 or more.');
            }
            if(!availabelDescription){
                setDescriptionError('The description\'s length should be 200 or more.');
            }
            if(!availabelContent){
                setContentError('The content\'s length should be 500 or more.');
            }
            if(!category){
                setCategoryError('Set a category to the Blog.');
            }
            //setError('Fill out all the fields || The title\'s length should be 90 or more, The description\'s length should be 255 or more, And the content\'s length should be 500 or more.');
        }
    };

    const editing = () => {
        const availabelTitle = title.replace(/\s/g, '').length >= 90;
        const availabelDescription = description.replace(/\s/g, '').length >= 200;
        const availabelContent = content.replace(/\s/g, '').length >= 500;
        const availabelData = (availabelTitle && availabelDescription && availabelContent && category);
        (availabelData) && edit();
    }

    return (
        <div className='content-container'>
            {error && <div>{error}</div>}
            <form className={isEdit ? 'blog-edit-form' : 'blog-new-form'} data-testid='form' onSubmit={onFormSubmit}>
                <div className='blog-form__lab-inp'>
                    <div>
                        <label htmlFor='title_new'>Title:</label>
                    </div>
                    <div>
                        <textarea
                          className='title-input'
                          id='title_new'
                          data-testid='title_input'
                          placeholder='Title (More than 90 characters)'
                          autoFocus
                          maxLength={120}
                          value={title}
                          onChange={onTitleChange}
                        />
                        {titleError && <p data-testid='title_error'>{titleError}</p>}
                    </div>
                </div>
                <div className='blog-form__lab-inp'>
                    <div>
                        <label htmlFor='description_new'>Description:</label>
                    </div>
                    <div>
                        <textarea
                          className='description-input'
                          id='description_new'
                          data-testid='description_input'
                          placeholder='Description (More than 255 characters)'
                          maxLength={300}
                          value={description}
                          onChange={onDescriptionChange}
                        />
                        {descriptionError && <p data-testid='description_error'>{descriptionError}</p>}
                    </div>
                </div>
                <div className='blog-form__lab-inp'>
                    <div>
                        <label htmlFor='content_new'>Content:</label>
                    </div>
                    <div>
                        <textarea
                          className='content-input'
                          id='content_new'
                          data-testid='content_input'
                          placeholder='Content (More than 500 characters)'
                          value={content}
                          onChange={onContentChange}
                        />
                        {contentError && <p data-testid='content_error'>{contentError}</p>}
                    </div>
                </div>
                <div className='blog-form__lab-inp'>
                    <div>
                        <label htmlFor='category_new'>Category:</label>
                    </div>
                    <div>
                        <select
                            className='category-selector'
                            id='category_new'
                            data-testid='category_select'
                            defaultValue={category ? category : ''}
                            onChange={onCategoryChange}
                        >
                            {
                                isEdit ?
                                <option disabled value={category}>{category}</option> :
                                <option disabled value=''>--Select Category--</option>
                            }
                            {
                                categories.map((category) => <option key={category} value={category}>{category}</option>)
                            }
                        </select>
                        {categoryError && <p data-testid='category_error'>{categoryError}</p>}
                    </div>
                </div>
                <div className='blog-form__button'>
                    {
                        isEdit ?
                        <button data-testid='edit_create_button' onClick={editing}>Edit Blog</button> :
                        <button data-testid='edit_create_button'>Create Blog</button>
                    }
                </div>
            </form>
        </div>
    );
};

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    profileList: state.userPage,
    savedBlogs: state.savedBlogs.blogs,
    categories: state.categories,
});

const mapDispatchToProps = (dispatch) => ({
    startAddBlog: (blog) => dispatch(startAddBlog(blog)),
    addBlogToProfile: (blog, id) => dispatch(addBlogToProfile(blog, id)),
    startEditBlog: (bid, updates) => dispatch(startEditBlog(bid, updates)),
    editBlogFromBP: (bid, updates) => dispatch(editBlogFromBP(bid, updates)),
    editBlogFromSB: (bid, updates) => dispatch(editBlogFromSB(bid, updates)),
    editBlogFromUP: (bid, updates) => dispatch(editBlogFromUP(bid, updates)),
});

export default connect(mapStateToProps, mapDispatchToProps)(NewBlog);