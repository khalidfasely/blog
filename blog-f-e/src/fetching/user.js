export default () => {
    return fetch('/data/user')
    .then(res => res.json())
    .then(result => {
        //console.log(result.user);
        if (result.user !== "AnonymousUser") {
            return result;
        } else {
            return undefined;
        }
    })
    .catch(er => console.log(er));
};