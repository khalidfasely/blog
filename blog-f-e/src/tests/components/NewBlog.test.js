import React from "react";
import { fireEvent, render } from '@testing-library/react';
import { user1 } from '../fixtures/user';
import { blog1 } from '../fixtures/blog';
import { NewBlog } from "../../components/NewBlog";

let props;

beforeEach(() => {
    props = {
        uname: 'UserAdmin',
        profileList: [user1],
        isEdit: false,
        blog: undefined,
        edit: jest.fn(),
        savedBlogs: ['1'],
        categories: ['Web', 'Programming', 'React']
    }
});

test('Should render NewBlog(mode: new blog) component correctly', () => {
    const { asFragment, getByTestId } = render(<NewBlog {...props} />);
    expect(asFragment()).toMatchSnapshot();

    const titleInputEl = getByTestId('title_input');
    expect(titleInputEl.value).toBe('');

    const descriptionInputEl = getByTestId('description_input');
    expect(descriptionInputEl.value).toBe('');

    const contentInputEl = getByTestId('content_input');
    expect(contentInputEl.value).toBe('');

    const categorySelectEl = getByTestId('category_select');
    expect(categorySelectEl.value).toBe('');

    const buttonEl = getByTestId('edit_create_button');
    expect(buttonEl.textContent).toBe('Create Blog');
});

test('Should render NewBlog(mode: editing) component correctly', () => {
    const { asFragment, getByTestId } = render(<NewBlog {...{...props, isEdit: true, blog: blog1}} />);
    expect(asFragment()).toMatchSnapshot();

    const titleInputEl = getByTestId('title_input');
    expect(titleInputEl.value).toBe(blog1.title);

    const descriptionInputEl = getByTestId('description_input');
    expect(descriptionInputEl.value).toBe(blog1.description);

    const contentInputEl = getByTestId('content_input');
    expect(contentInputEl.value).toBe(blog1.content);

    const categorySelectEl = getByTestId('category_select');
    expect(categorySelectEl.value).toBe(blog1.category);

    const buttonEl = getByTestId('edit_create_button');
    expect(buttonEl.textContent).toBe('Edit Blog');
});

test('Should change inputs on NewBlog(mode: new blog) work correctly', () => {
    const { getByTestId } = render(<NewBlog {...props} />);

    const titleInputEl = getByTestId('title_input');
    fireEvent.change(titleInputEl, {
        target: {
            value: 'title'
        }
    });
    expect(titleInputEl.value).toBe('title');

    const descriptionInputEl = getByTestId('description_input');
    fireEvent.change(descriptionInputEl, {
        target: {
            value: 'description'
        }
    });
    expect(descriptionInputEl.value).toBe('description');

    const contentInputEl = getByTestId('content_input');
    fireEvent.change(contentInputEl, {
        target: {
            value: 'content'
        }
    });
    expect(contentInputEl.value).toBe('content');

    const categorySelectEl = getByTestId('category_select');
    fireEvent.change(categorySelectEl, {
        target: {
            value: 'Web'
        }
    });
    expect(categorySelectEl.value).toBe('Web');
});

test('Should change inputs on NewBlog(mode: editing) work correctly', () => {
    const { getByTestId } = render(<NewBlog {...{...props, isEdit: true, blog: blog1}} />);

    const titleInputEl = getByTestId('title_input');
    fireEvent.change(titleInputEl, {
        target: {
            value: 'edited title'
        }
    });
    expect(titleInputEl.value).toBe('edited title');

    const descriptionInputEl = getByTestId('description_input');
    fireEvent.change(descriptionInputEl, {
        target: {
            value: 'edited description'
        }
    });
    expect(descriptionInputEl.value).toBe('edited description');

    const contentInputEl = getByTestId('content_input');
    fireEvent.change(contentInputEl, {
        target: {
            value: 'edited content'
        }
    });
    expect(contentInputEl.value).toBe('edited content');

    const categorySelectEl = getByTestId('category_select');
    fireEvent.change(categorySelectEl, {
        target: {
            value: 'React'
        }
    });
    expect(categorySelectEl.value).toBe('React');
});

test('Should fire onsubmit on NewBlog(mode: new blog) with errors correctly', () => {
    const { getByTestId } = render(<NewBlog {...props} />);
    const formEl = getByTestId('form');

    const titleInputEl = getByTestId('title_input');
    fireEvent.change(titleInputEl, {
        target: {
            value: 'title'
        }
    });
    //fireEvent.change(titleInputEl, {
        //    target: {
    //        value: '12345678910 12345678910 12345678910 12345678910 12345678910 12345678910 12345678910 12345678910 12345678910'
    //    }
    //});
    //fireEvent.submit(formEl);
    //expect(getByText('The title\'s length should be 90 or more.')).not.toBeInTheDocument();
    
    const descriptionInputEl = getByTestId('description_input');
    fireEvent.change(descriptionInputEl, {
        target: {
            value: 'description'
        }
    });
    
    const contentInputEl = getByTestId('content_input');
    fireEvent.change(contentInputEl, {
        target: {
            value: 'content'
        }
    });
    
    const categorySelectEl = getByTestId('category_select');
    fireEvent.change(categorySelectEl, {
        target: {
            value: ''
        }
    });
    fireEvent.submit(formEl);
    
    const errorTitleEl = getByTestId('title_error');
    expect(errorTitleEl.textContent).toBe('The title\'s length should be 90 or more.');
    
    const errorDescriptionEl = getByTestId('description_error');
    expect(errorDescriptionEl.textContent).toBe('The description\'s length should be 200 or more.');
    
    const errorContentEl = getByTestId('content_error');
    expect(errorContentEl.textContent).toBe('The content\'s length should be 500 or more.');
    
    const errorCategoryEl = getByTestId('category_error');
    expect(errorCategoryEl.textContent).toBe('Set a category to the Blog.');
});

test('Should fire onsubmit on NewBlog(mode: editing) with errors correctly', () => {
    const { getByTestId } = render(<NewBlog {...{...props, isEdit: true, blog: blog1}} />);
    const formEl = getByTestId('form');

    const titleInputEl = getByTestId('title_input');
    fireEvent.change(titleInputEl, {
        target: {
            value: 'title'
        }
    });
    
    const descriptionInputEl = getByTestId('description_input');
    fireEvent.change(descriptionInputEl, {
        target: {
            value: 'description'
        }
    });
    
    const contentInputEl = getByTestId('content_input');
    fireEvent.change(contentInputEl, {
        target: {
            value: 'content'
        }
    });
    
    const categorySelectEl = getByTestId('category_select');
    fireEvent.change(categorySelectEl, {
        target: {
            value: ''
        }
    });

    fireEvent.submit(formEl);

    const errorTitleEl = getByTestId('title_error');
    expect(errorTitleEl.textContent).toBe('The title\'s length should be 90 or more.');

    const errorDescriptionEl = getByTestId('description_error');
    expect(errorDescriptionEl.textContent).toBe('The description\'s length should be 200 or more.');

    const errorContentEl = getByTestId('content_error');
    expect(errorContentEl.textContent).toBe('The content\'s length should be 500 or more.');

    const errorCategoryEl = getByTestId('category_error');
    expect(errorCategoryEl.textContent).toBe('Set a category to the Blog.');
});

test('Should edit() call correctly', () => {
    const { getByTestId } = render(<NewBlog {...{...props, isEdit: true, blog: blog1}} />);

    const buttonEditEl = getByTestId('edit_create_button');
    expect(buttonEditEl.textContent).toBe('Edit Blog');

    expect(props.edit).toHaveBeenCalledTimes(0);
    fireEvent.click(buttonEditEl);
    //Because the data is not available(availableData === false)
    expect(props.edit).toHaveBeenCalledTimes(0);
});