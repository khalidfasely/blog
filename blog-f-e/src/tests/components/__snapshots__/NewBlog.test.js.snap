// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render NewBlog(mode: editing) component correctly 1`] = `
<DocumentFragment>
  <div
    class="content-container"
  >
    <form
      class="blog-edit-form"
      data-testid="form"
    >
      <div
        class="blog-form__lab-inp"
      >
        <div>
          <label
            for="title_new"
          >
            Title:
          </label>
        </div>
        <div>
          <textarea
            class="title-input"
            data-testid="title_input"
            id="title_new"
            maxlength="120"
            placeholder="Title (More than 90 characters)"
          >
            title2
          </textarea>
        </div>
      </div>
      <div
        class="blog-form__lab-inp"
      >
        <div>
          <label
            for="description_new"
          >
            Description:
          </label>
        </div>
        <div>
          <textarea
            class="description-input"
            data-testid="description_input"
            id="description_new"
            maxlength="300"
            placeholder="Description (More than 255 characters)"
          >
            description
          </textarea>
        </div>
      </div>
      <div
        class="blog-form__lab-inp"
      >
        <div>
          <label
            for="content_new"
          >
            Content:
          </label>
        </div>
        <div>
          <textarea
            class="content-input"
            data-testid="content_input"
            id="content_new"
            placeholder="Content (More than 500 characters)"
          >
            content
          </textarea>
        </div>
      </div>
      <div
        class="blog-form__lab-inp"
      >
        <div>
          <label
            for="category_new"
          >
            Category:
          </label>
        </div>
        <div>
          <select
            class="category-selector"
            data-testid="category_select"
            id="category_new"
          >
            <option
              disabled=""
              selected=""
              value="Django"
            >
              Django
            </option>
            <option
              value="Web"
            >
              Web
            </option>
            <option
              value="Programming"
            >
              Programming
            </option>
            <option
              value="React"
            >
              React
            </option>
          </select>
        </div>
      </div>
      <div
        class="blog-form__button"
      >
        <button
          data-testid="edit_create_button"
        >
          Edit Blog
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Should render NewBlog(mode: new blog) component correctly 1`] = `
<DocumentFragment>
  <div
    class="content-container"
  >
    <form
      class="blog-new-form"
      data-testid="form"
    >
      <div
        class="blog-form__lab-inp"
      >
        <div>
          <label
            for="title_new"
          >
            Title:
          </label>
        </div>
        <div>
          <textarea
            class="title-input"
            data-testid="title_input"
            id="title_new"
            maxlength="120"
            placeholder="Title (More than 90 characters)"
          />
        </div>
      </div>
      <div
        class="blog-form__lab-inp"
      >
        <div>
          <label
            for="description_new"
          >
            Description:
          </label>
        </div>
        <div>
          <textarea
            class="description-input"
            data-testid="description_input"
            id="description_new"
            maxlength="300"
            placeholder="Description (More than 255 characters)"
          />
        </div>
      </div>
      <div
        class="blog-form__lab-inp"
      >
        <div>
          <label
            for="content_new"
          >
            Content:
          </label>
        </div>
        <div>
          <textarea
            class="content-input"
            data-testid="content_input"
            id="content_new"
            placeholder="Content (More than 500 characters)"
          />
        </div>
      </div>
      <div
        class="blog-form__lab-inp"
      >
        <div>
          <label
            for="category_new"
          >
            Category:
          </label>
        </div>
        <div>
          <select
            class="category-selector"
            data-testid="category_select"
            id="category_new"
          >
            <option
              disabled=""
              selected=""
              value=""
            >
              --Select Category--
            </option>
            <option
              value="Web"
            >
              Web
            </option>
            <option
              value="Programming"
            >
              Programming
            </option>
            <option
              value="React"
            >
              React
            </option>
          </select>
        </div>
      </div>
      <div
        class="blog-form__button"
      >
        <button
          data-testid="edit_create_button"
        >
          Create Blog
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
